### ISSUE TITLE
Podcast "Unable to retrieve asset details from Sable(CDN), fallback to CDS"

### ISSUE THREAD

**Issue by rmcrackan** [rmcrackan/Libation#122]
@mkb79 , @Mbucari 

Dear heavens podcast episodes are messed up. Below is for [this podcast](https://www.audible.com/pd/Podcast/B08K56QKL3). Any ideas on this one? The error is:

```
{
  "error_code": "000307",
  "message": "Unable to retrieve asset details from Sable(CDN), fallback to CDS for customerId:..., marketplaceId:AF2M0KC94RCEA, asin:B099QTVZZR, acr:null, skuLite:PC_800I_067097, aaaClientId:ApolloEnv:AudibleApiExternalRouterService/NA/ACLS/Prod"
}
```

When I make a call to https://cds.audible.... it returns "No Library Record Found for user."

Even if I add the specific episode to your library via my phone, a desktop library book api call returns "ASIN B09GW57QYW not present in customer library."

@Mbucari 

This brings up an unexpected issue. If there are episodes or similar which do need to decode the old way (pre-license files), can AAXClean handle this?

**Comment by Mbucari** [comment]
Things like this were so much easier to debug when the Libation solution included the AudibleApi source!

I seem to remember seeing that error a couple months ago while debugging some of  wtanksleyjr's errors. IIRC, it happened when I attempted to download a book that was no longer in my account's library, but was still in Libation's library.  I just tried reproducing it now by adding an audible plus title, importing into Libation, removing that title from my audible library, then attempting to download the title. Surprisingly, that action is now permissible! I was able to download an audible plus title not in my library!

> If there are episodes or similar which do need to decode the old way (pre-license files), can AAXClean handle this?

Yes. Just use a different SetDecryptionKey overload to set the activation_bytes. I assume you ask because you think the old "workaround" download method my work for podcasts?

**Comment by rmcrackan** [comment]
> Things like this were so much easier to debug when the Libation solution included the AudibleApi source!

What I do for debugging is: create a new solution folder "__nuget". rt-clk it: Add existing project: AudibleApi. InternalUtilities adds reference to that. You don't even need to remove the nuget pkg as long as the pkg and the code base are up to date. Now it will debug into the project. When done: remove the AudibleApi project from the solution and it will automatically get removed from InternalUtilities too.

Are you able to download anything from this podcast?  
https://www.audible.com/pd/Podcast/B08K56QKL3  
I spent hours on these today and made exactly zero progress

> I assume you ask because you think the old "workaround" download method my work for podcasts?

Basically yes but I'm not for sure. The "fallback to CDS" error message made me wonder though.

**Comment by Mbucari** [comment]
I think I figured it out. I used fiddler to inspect the requests audible is making when it plays a podcast from the web browser.  It does make a call to the /1.0/content/%{asin}/licenserequest api endpoint, but it set these parameters:

```JSON
{
   "response_groups":"chapter_info,content_reference,last_position_heard",
   "chapter_titles_type":"Tree",
   "supported_drm_types":[
      "Hls",
      "Mpeg",
      "Dash",
      "Widevine"
   ],
   "use_adaptive_bit_rate":true,
   "consumption_type":"Streaming"
}
```
Inside GetDownloadLicenseAsync, I changed the request body to:

```C#
            var body = new JObject
            {
                { "consumption_type", "Streaming" },
                { "drm_type", "Mpeg" },
                { "quality", quality.ToString() },
                { "response_groups", "last_position_heard,content_reference,chapter_info"}
            };
``` 
And I was able to get a download license (though it threw an exception while parsing because the parser is expecting an encrypted voucher, and the podcast isn't encrypted).

The response for that request is:

```JSON
{"content_license": {
  "access_expiry_date": "REDACTED",
  "acr": "CR!3EVB7MCZTD08S9XBMASS8899SWEC",
  "asin": "B095N15ZLM",
  "content_metadata": {
    "chapter_info": {
      "brandIntroDurationMs": 0,
      "brandOutroDurationMs": 0,
      "chapters": [
        {
          "length_ms": 720000,
          "start_offset_ms": 0,
          "start_offset_sec": 0,
          "title": "Wizards of High Sorcery"
        }
      ],
      "is_accurate": false,
      "runtime_length_ms": 720000,
      "runtime_length_sec": 720
    },
    "content_reference": {
      "acr": "CR!3EVB7MCZTD08S9XBMASS8899SWEC",
      "asin": "B095N15ZLM",
      "content_format": "MPEG",
      "content_size_in_bytes": 17287761,
      "file_version": "1",
      "marketplace": "AF2M0KC94RCEA",
      "sku": "PC_800F_305820",
      "tempo": "1.0"
    },
    "last_position_heard": {
      "status": "DoesNotExist"
    }
  },
  "drm_type": "Mpeg",
  "license_id": "REDACTED",
  "license_response": "https://media.blubrry.com/dragonlancesaga/dragonlance-saga.com/audio/wizards-of-high-sorcery.mp3",
  "message": "Customer [REDACTED] has rights to asin [B095N15ZLM] for AYCL",
  "request_id": "REDACTED",
  "requires_ad_supported_playback": false,
  "status_code": "Granted"
}}
```

**Comment by Mbucari** [comment]
I tried replacing ``` { "drm_type", "Mpeg" },``` to ```{ "supported_drm_types", new JArray{ "Mpeg", "Adrm" } } ```, but the api returns a "Bad Request" error. It will not work if you include "Adrm" as a drm type. Also, I tried to get a license using ```"consumption_type":"Download"``` and AudibleApi successfully parsed the response! So all we need to do is figure out how to tell ahead of time if a podcast has drm or not, and pass "Mpeg" drm_type for those and "Adrm" for everything else.

**Comment by rmcrackan** [comment]
Damn, I wish I'd seen this last edit before I meticulously went through the json. lol

I think the answer might be that unencrypted books have

`
    "available_codecs": null,
`

**Comment by Mbucari** [comment]
> Damn, I wish I'd seen this last edit before I meticulously went through the json. lol

Yeah, there really should be notifications for edits lol.

I created a small PR for AudibleApi.  Now you just have to decide how to handle these no-drm files. The files from that podcast are mp3s, so AAXClean can't read metadata from them. that means incomplete nfos, unless you bring back taglib-sharm to handle those files.  The content license contains the ChapterInfo structure, so you can create a cue from that.

**Comment by Mbucari** [comment]
Side note: apparently visual studio can't correctly hot reload changes to JObject literals.

**Comment by rmcrackan** [comment]
I accepted the merge. In Libation I got rid of some of the null/empty checks for DownloadLicense since unencrypted files don't have IV or Key. I also deleted NFO files. Not adding value. If this is something worth keeping, it should be stored in a real db.

* DownloadDecryptBook's format switch feels wrong now
* In AaxcDownloadConverter, new AaxFile throws a NRE. I assume because of these blank key and IV

**Comment by rmcrackan** [comment]
I pushed my changes so you can see them too

**Comment by Mbucari** [comment]
>     * In AaxcDownloadConverter, new AaxFile throws a NRE. I assume because of these blank key and IV
 
No, it's because that podcast is an mp3 file, and AAXClean can only read mp4. So if the source is MP3, then Step 1 won't be able to retrieve metadata.

>  * DownloadDecryptBook's format switch feels wrong now

Agreed. Now, we have to determine the source format (Aax or Mp3), and decide the output format based on the input format + user settings.  Since it's now clear that we can get source files in mp3 format, the user setting needs to be a choice between "Original Format (Lossless)" and "Always Mp3 (transcode if needed)".

 So far I have not seen an unencrypted Mp4 file, but they may exist.

**Comment by Mbucari** [comment]
Right now I'm torn on how to go about this. There are three options I'm considering: 
1. Modify AaxcDownloadConverter to handle unencrypted downloads. But that's not really its job, and that solution seems kind of hacky. There would be so many special cases and it would really complicate the flow.
2. Move api.GetDownloadLicenseAsync() into DownloadDecryptBook.ProcessAsync().  If the DrmType is Adrm, we call aaxToM4bConverterDecryptAsync() as usual.  If DrmType is Mpeg, we handle it in a different method with a different class.  Possibly move a lot of AaxcDownloadConverter's functionality into an abstract class and make new derived types to handle Drm and non-drm books.
3. Go WAY upstream and get the download license inside ProcessorAutomationController.  Then, based on the the DrmType, either spawn an AudioDecryptForm with a DownloadDecryptBook, or a DownloadForm with a DownloadFile.

I think I like option 2 the most.

**Comment by rmcrackan** [comment]
3 is a non-starter. This is part of biz logic flow which doesn't belong in the controller/UI-wire-up. 1 might work but I agree that 2 is much better.

**Comment by rmcrackan** [comment]
Solved in #123
